### PR TITLE
Implement device driver for BNO055 9DOF sensor

### PR DESCRIPTION
**Pull request by Achilleshiel**
### Contribution description

Driver for the BOSCH BNO055 IMU-sensor. This sensor has an internal MCU that does a lot of automatic sensor fusion magic for you.
I added some conversion functions for the float values as some sensor fusion outputs are only usefull when they are specified as floats. However, one can still read the raw integer data. And depending on the setting it can easily be scaled to SI-units.

The driver has read functions for:
- Euler Angle rotation
- Quaternion rotation
- Linear Acceleration (removed the gravitational acceleration from data output)
- Gravity Vector (outputs only gravitational acceleration)
- Acceleration
- Angular rate
- Magnetic field strength

It can also select units for:
- Acceleration: m/s^2 or mg (9.81mm/s^2)
- Magnetic Field strength: μT
- Euler Angles: Degrees or Radians
- Angular Rate: Degree/s or Radians/s

### Testing procedure
I made a testing-program that reads and prints all the sensor values.

### Issues/PRs references
--

### CODE REVIEW

**Comment by fjmolinas** [comment]
Hi @Achilleshiel thanks a lot for the contribution and welcome to RIOT!

I took a quick glimpse at the code, looks good, I'll take a more detailed look tomorrow. Travis pointed out a couple of nitpicks though. Also AFAIK `SAUL` integration seems to be missing in this PR, there is some general information in http://riot-os.org/api/driver-guide.html. Do you think you would be able to implement that? Let me know if you need some help with that or figuring out travis!

BTW do you have some test output you can show? I don't have the driver to test it but I can trust your test results!

**Comment by Achilleshiel** [comment]
Hi @fjmolinas,

I think I have statisfied Travis now. Except that I have to squash this PR at some point =)
I will have a look into the SAUL-part in the weekend. If I have any questions I will let you know.

The output of my test:
```
2020-09-22 20:13:21,549 # main(): This is RIOT! (Version: 2020.10-devel-1434-gace4-bno055)
2020-09-22 20:13:22,551 # Test for BNO055 device
2020-09-22 20:13:22,556 # Initialize BNO055
2020-09-22 20:13:22,557 # Initialized succesfull
2020-09-22 20:13:23,571 # Quaternions: 0.872803 + -0.274353i + -0.403625j +-0.002441k
2020-09-22 20:13:23,572 # Euler rotation: Heading: 0.006, Roll 0.798, Pitch 0.751
2020-09-22 20:13:23,577 # Gravitation Vector: x: 6.92, y: -4.67, z:5.13
2020-09-22 20:13:23,583 # Linear Acceleration: x: -0.04, y: 0.12, z:0.13
2020-09-22 20:13:23,588 # Magnetometer: x: -20.50, y: 28.69, z:-4.69
2020-09-22 20:13:23,589 # Accelerometer: x: 6.85, y: -4.54, z:5.28
2020-09-22 20:13:23,601 # Gyroscope: x: 0.01, y: 0.05, z:-0.02
2020-09-22 20:13:24,603 # Quaternions: 0.874329 + -0.271790i + -0.402161j +-0.002686k
2020-09-22 20:13:24,609 # Euler rotation: Heading: 0.006, Roll 0.796, Pitch 0.743
2020-09-22 20:13:24,615 # Gravitation Vector: x: 6.90, y: -4.63, z:5.18
2020-09-22 20:13:24,621 # Linear Acceleration: x: -0.14, y: 0.18, z:0.03
2020-09-22 20:13:24,633 # Magnetometer: x: -21.56, y: 29.06, z:-4.69
2020-09-22 20:13:24,633 # Accelerometer: x: 6.83, y: -4.49, z:5.32
2020-09-22 20:13:24,638 # Gyroscope: x: -0.01, y: 0.03, z:-0.02
2020-09-22 20:13:25,122 # Exiting Pyterm
```

File: tests/driver_bno055/main.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+{
+    xtimer_sleep(1);
+    printf("Test for BNO055 device\n");
+    int result;
+    bno055_t dev;
+
+    printf("Initialize BNO055\n");
+    result = bno055_init(&dev, &bno055_params[0]);
+    if (result != BNO055_OK) {
+        printf("[Error] Failed to connect with the i2c device\n");
```
Comment (aabadie) [inline tests/driver_bno055/main.c:37]:
```suggestion
        printf("[Error] Failed to initialize the i2c device\n");
```

File: tests/driver_bno055/main.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    printf("Initialized successful\n");
```
Comment (aabadie) [inline tests/driver_bno055/main.c:40]:
```suggestion
    printf("Initialization successful\n");
```

**Review by aabadie — COMMENTED** [review]
Looks almost good.

We recently updated the device driver implementation guide to encourage the use of errno codes instead of custom return codes. I won't insist on this point but if you feel like it and want to update this part, that would be nice.
Another thing that would be nice to add is a basic README for the test application.

File: drivers/include/bno055.h [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+/**
+ * @brief   Status and error return codes
+ */
+enum {
+    BNO055_OK       = 0,      /**< exit without error */
+    BNO055_NOBUS    = ENXIO,  /**< cannot connect to module on i2c bus */
+    BNO055_NODEV    = ENODEV, /**< cannot read any data from module */
+    BNO055_NORW     = EIO,    /**< cannot read data from module */
+    BNO055_NOTREADY = EBUSY,  /**< no new data ready for reading */
+};
```
Comment (aabadie) [inline drivers/include/bno055.h:113]: This whole enum should be removed and the errno codes directly returned in the implementation. The doxygen documentation also needs to be adapted.

File: drivers/include/bno055.h [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_bno055 BNO055 9-axis sensor
+ * @ingroup     drivers_sensors
+ * @ingroup     drivers_saul
+ * @brief       Device driver for the Bosch BNO055 9-axis sensor
+ *
+ * This driver provides @ref drivers_saul capabilities.
```
Comment (benpicco) [inline drivers/include/bno055.h:15]: You didn't provide a `bno055_saul.c ` for that :wink:

File: drivers/bno055/include/bno055_internal.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name    BNO055 chip id registers and who am I values
+ * @{
+ */
+#define BNO055_REG_CHIPID          (0x00U)  /**> chip id register */
```
Comment (aabadie) [inline drivers/bno055/include/bno055_internal.h:30]:
```suggestion
#define BNO055_REG_CHIPID          (0x00U)  /**< chip id register */
```

This doxygen issue is present all over the place.
In current master:
```
$ git grep "/\*\*>" | wc -l
0
git grep "/\*\*<" | wc -l
373126
```

**Comment by fjmolinas** [comment]
@Achilleshiel there seems to be very little missing here, any chance you can take a look at the missing comments?

**Pull request closed without merging**